src/server/ua_file.c

```diff
@@ -0,0 +1,419 @@
+setFileMethodCallbacks(UA_Server *server, const UA_NodeId fileObjectNodeId)
+{
+    UA_StatusCode result = UA_STATUSCODE_GOOD;
+    UA_NodeId openNodeId;
+    UA_NodeId readNodeId;
+    UA_NodeId setPositionNodeId;
+    UA_NodeId getPositionNodeId;
+    UA_NodeId writeNodeId;
+    UA_NodeId closeNodeId;
+
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldOpenQN, &openNodeId);
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldReadQN, &readNodeId);
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldSetPositionQN, &setPositionNodeId);
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldGetPositionQN, &getPositionNodeId);
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldWriteQN, &writeNodeId);
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldCloseQN, &closeNodeId);
+
+    result |= UA_Server_setMethodNode_callback(server, openNodeId, UA_Server_openFileCallback);
+    result |= UA_Server_setMethodNode_callback(server, readNodeId, UA_Server_readFileCallback);
+    result |= UA_Server_setMethodNode_callback(server, setPositionNodeId, UA_Server_setPositionFileCallback);
+    result |= UA_Server_setMethodNode_callback(server, getPositionNodeId, UA_Server_getPositionFileCallback);
+    result |= UA_Server_setMethodNode_callback(server, writeNodeId, UA_Server_writeFileCallback);
+    result |= UA_Server_setMethodNode_callback(server, closeNodeId, UA_Server_closeFileCallback);
+
+    return result;
+}
+
+UA_StatusCode
+setFileTypeInfo(UA_Server *server, const UA_NodeId fileTypeObjectNodeId,
+                          UA_String filePath)
+{
+    UA_FileType *fileTypeObject = (UA_FileType*) UA_malloc(sizeof(UA_FileType));
+    UA_NodeId_copy(&fileTypeObjectNodeId, &fileTypeObject->fileTypeObjectNodeId);
+    UA_String_copy(&filePath, &fileTypeObject->filePath);
+    fileTypeObject->openCount = 0;
+    fileTypeObject->fileSize = 0;
+    fileTypeObject->writable = UA_TRUE;
+    fileTypeObject->userWritable = UA_TRUE;
+    setFileObjectNodeProperties(server, fileTypeObject);
+    LIST_INIT(&fileTypeObject->fileInfo);
+    LIST_INSERT_HEAD(&server->fileObjects, fileTypeObject, listEntry);
+    return UA_STATUSCODE_GOOD;
+}
+
+UA_StatusCode 
+UA_Server_addFile(UA_Server *server, const UA_NodeId requestedNewNodeId,
+                  const UA_NodeId parentNodeId,
+                  const UA_NodeId referenceTypeId,
+                  const UA_QualifiedName browseName,
+                  const UA_ObjectAttributes attr,
+                  const UA_String filePath,
+                  void *nodeContext, UA_NodeId *outNewNodeId)
+{
+    UA_StatusCode result = UA_Server_addObjectNode(server, requestedNewNodeId,
+                            parentNodeId, referenceTypeId, browseName,
+                            UA_NODEID_NUMERIC(0, UA_NS0ID_FILETYPE),
+                            attr, nodeContext, outNewNodeId);
+    
+    UA_file *sampleFile = UA_file_open((char*)filePath.data, "w");
```

Comment:
Error-check everywhere.
This should fail when the file does not exist.
Use proper logging,

Reply:
This can only log an error and return the invalid `UA_file*` to the caller.